File: lab3/part1/task4/plugins/Parrot.java
package task4.plugins;

import task4.Animal;

public final class Parrot extends Animal {

    private final String name;

    public Parrot(String name) {
        this.name = name;
    }

    @Override
    public String name() {
        return this.name;
    }

    @Override
    public String greet() {
        return "yaaaaarrrr!";
    }

    @Override
    public String menu() {
        return "seeds";
    }
}

File: lab3/part1/parrot.c
#include "animal.h"
#include <stdlib.h>

char const* name(struct Animal* this) {
    return this->name;
}

char const* greet(void) {
    return "yaaarrrr!";
}

char const* menu(void) {
    return "brazillian nuts";
}

PTRFUN table[] = {
    name,
    greet,
    menu
};

void construct(struct Animal* memory, const char* name) {
    memory->name = name;
    memory->vtable = table;
}

struct Animal* create(char const* name) {
    struct Animal* parrot = (struct Animal*) malloc(sizeof(struct Animal));
    construct(parrot, name);
    return parrot;
}
